### ISSUE TITLE
New rule suggestion: Require `await` for resolving promises.

### ISSUE THREAD

**Issue by garyking** [typescript-eslint/typescript-eslint#1203]
**Repro**

`.eslintrc.js`:

```js
module.exports = {
  extends: ['plugin:@typescript-eslint/recommended'],
  parserOptions: {
    project: './tsconfig.json',
  },
  rules: {
    '@typescript-eslint/no-floating-promises': 'error',
    '@typescript-eslint/no-misused-promises': 'error',
    '@typescript-eslint/require-await': 'error',
  },
};
```

`src/index.ts`:

```TS
import globby from 'globby';

// Await function.
async function getFiles(): Promise<string[]> {
  const files = await globby(__dirname);

  return files;
}

async function main(): Promise<void> {
  // Should say that `getFiles` needs to be called with `await`, but it doesn't
  // say anything.
  const files = getFiles();

  console.log(files);
}

main();
```

**Expected Result**

We should have a new rule, where it reports that any functions that return a promise, should be called with `await`, to resolve the promise.

**Actual Result**

We don't get any errors on this. Even with the rules I've got in my sample config above.

**Comment by bradzacher** [comment]
Rules that solve this in some way:

- [`no-floating-promises`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-floating-promises.md) will ensure that every promise has an `await` (or a `.then`/`.catch`).
- [`require-await`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/require-await.md) will ensure that every `async` function has at least one `await` within it (or that it returns something of type `Promise`).
- [`no-misused-promises`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-misused-promises.md) will make sure you aren't doing some common incorrect things with promises.

**Comment by garyking** [comment]
@bradzacher As indicated above, I've got all 3 rules enabled, and none are reporting on the line that's missing an `await`, though.

**Comment by garyking** [comment]
@bradzacher A simple piece of code like this should warn to use `await` in the `no-floating-promises` rule, but it doesn't even do that:

```ts
async function main(): Promise<void> {
  const promise = Promise.resolve('value');

  console.log(promise);
}

main();
```

**Comment by bradzacher** [comment]
I was wrong, sorry - `no-floating-promise` won't catch anything here, because the promise isn't floating - it's assigned to a variable.
If you had
```ts
async function main(): Promise<void> {
  Promise.resolve('value');
}
```
Then `no-floating-promise` should error on the promise.

----

`require-await` won't error on the line that's missing an `await`, it will error on the function itself if it's marked `async`, but contains no awaits (or doesn't return a promise).

---

`no-misused-promises` will check for things like
```ts
if (returnsPromise()) {}
```

But it won't catch an unawaited promise

---

The problem is that this is a pretty hard problem to lint for.
`require-await` is a good start, as it will error for some simple cases, but will obv miss other cases.

The problem that's hard to solve is that you have to do control-flow analysis to figure out if a promise is not returned or not awaited.

This is in of itself, perfectly valid code:
```ts
async function foo(): Promise<void> {
  const x = await returnsPromise();
  const y = returnsPromise();
}
```
though it's obvious upon inspection that the intention was probably to await the second promise.

But you can't just error there, because what happens if you're intentionally doing something concurrently?

```ts
async function foo(): Promise<void> {
  const x = returnsPromise();
  const y = returnsPromise();

  await x;
  await y;
}
```
the un-awaited promises are 100% valid in this case, right?

So we want to track every single promise to make sure it's awaited, but what about a case like this:

```ts
async function foo(): Promise<void> {
  const x = returnsPromise();
  const y = returnsPromise();

  await x;

  if (someCOndition) {
    return;
  }

  await y;
}
```

Do we error for y? we have to do control-flow analysis to attempt to figure out that `y` is not always awaited.

And then there's the problem of helper methods for doing things like:
```ts
async function foo(): Promise<void> {
  const x = returnsPromise();
  const y = returnsPromise();

  await Promise.all([x, y]);
}
```
techincally `x` and `y` aren't directly awaited, but how can we know that this function properly consumes and awaits the promises. Sure we can encode information about the builtin `Promise` static methods, but then custom functions will cause errors..

```ts
const Bluebird = require('bluebird');
async function foo(): Promise<void> {
  const x = returnsPromise();
  const y = returnsPromise();

  await Bluebird.race([x, y]);
}
```

and then things get more sticky if you're doing dynamic promises:

```ts
async function foo(): Promise<void> {
  const proms = [];
  for (let i = 0; i < 10; i += 1) {
    proms.push(returnsPromise(i));
  }
  await Promise.all(proms);
}
```

Sorry, I ranted here a lot - what I'm trying to get at is that it's actually really, really, really hard to do what you want in a way that won't false-positive or break willy-nilly.

**Comment by garyking** [comment]
Thanks, I read the whole thing. I get it now. 

As you can see, my idea for this rule came from a simple example, in which I figured that it would be nice to be warned when what I was dealing with was a promise, rather than the result of one. But I now see that there are too many cases where a promise is intentionally consumed without resolving it immediately.

I've written ESLint rules before, and typically I only notice later when the number of situations that they apply to is far more than I initially assumed.
